harness/determined/tensorboard/fetchers/s3.py

```diff
@@ -0,0 +1,87 @@
+import datetime
+import logging
+import os
+import urllib.parse
+from typing import Any, Dict, Generator, List, Tuple
+
+from .base import Fetcher
+
+logger = logging.getLogger(__name__)
+
+
+class S3Fetcher(Fetcher):
+    def __init__(
+        self, storage_config: Dict[str, Any], storage_paths: List[str], local_dir: str
+    ) -> None:
+        import boto3
+
+        from determined.common.storage import boto3_credential_manager
+
+        boto3_credential_manager.initialize_boto3_credential_providers()
+        self.s3 = boto3.resource(
+            "s3",
+            endpoint_url=storage_config.get("endpoint_url", None),
+            aws_access_key_id=storage_config.get("access_key", None),
+            aws_secret_access_key=storage_config.get("secret_key", None),
+        )
+        self.client = self.s3.meta.client
+        self.bucket_name = str(storage_config["bucket"])
+
+        self.local_dir = local_dir
+        self.storage_paths = storage_paths
+        self._file_records = {}  # type: Dict[str, datetime.datetime]
+
+    def _find_keys(self, prefix: str) -> Generator[Tuple[str, datetime.datetime], None, None]:
+        logger.debug(f"Listing keys in bucket '{self.bucket_name}' with prefix '{prefix}'")
+
+        prefix = urllib.parse.urlparse(prefix).path.lstrip("/")
+
+        list_objs_args = {}  # type: Dict[str, Any]
+        while True:
```

Comment:
could you please use boto paginator machinery instead of doing it manually: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

Reply:
Oh nice! I don't think this existed the last time I wrote code for boto3.